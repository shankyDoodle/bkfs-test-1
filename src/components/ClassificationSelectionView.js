import React from 'react';
import _ from 'lodash'

import MSSView from "../libraries/MSS/MSSView";
import {bindActionCreators} from "redux";
import * as myActions from "../actions";
import {connect} from "react-redux";
import {Button, Divider, Popover} from 'antd';

import {dropdownTypes, screenNames} from "../constants/appConstants";

class ClassificationSelectionView extends React.Component {

    createDropDownListModel(data) {
        let list = []
        for (let key in data) {
            let oData = data[key];
            list.push({id: oData.id, label: oData.label})
        }
        return list;
    }

    handleClassificationDropDownOnBlur=(dropdownButtonType, selectedItems)=>{
        this.props.dispatch(myActions.handleClassificationDropDownOnBlur(dropdownButtonType, selectedItems));
    }

    handleCreateButtonClicked=()=>{
        let selectedCustomerIds = this.props.selectedCustomers
        let selectedDocumentTypeIds = this.props.selectedDocuments;
        this.props.dispatch(myActions.handleClassificationCreateButtonCLickedServerCall(selectedCustomerIds, selectedDocumentTypeIds));
    }

    handleNewCustomerAdded=(sCustomerName)=>{
        this.props.dispatch(myActions.handleNewCustomerAdded(sCustomerName));
    }

    handleNewDocumentTypeAdded=(sDocTypeName)=>{
        this.props.dispatch(myActions.handleNewDocumentTypeAdded(sDocTypeName));
    }

    getCustomerDropdownView(){
        let aDropDownListModel = this.createDropDownListModel(this.props.customerList)
        let isDisabled = this.props.customerData.hasOwnProperty("clonedObject")
        return <MSSView
            key={dropdownTypes.CUSTOMER_NAMES}
            label={"Customer Names"}
            childElements={aDropDownListModel}
            onChange={this.handleClassificationDropDownOnBlur.bind(this, dropdownTypes.CUSTOMER_NAMES)}
            isMultiple={true}
            allowClear={true}
            disabled={isDisabled}
            selected={this.props.selectedCustomers}
            onEnterPress={this.handleNewCustomerAdded}
            selectAll={false}/>
    }

    getDocumentDropdownView(){
        let aDropDownListModel = this.createDropDownListModel(this.props.documentTypes)
        let isDisabled = this.props.customerData.hasOwnProperty("clonedObject")
        return <MSSView
            key={dropdownTypes.DOCUMENT_TYPES}
            label={"Document Types"}
            childElements={aDropDownListModel}
            onChange={this.handleClassificationDropDownOnBlur.bind(this, dropdownTypes.DOCUMENT_TYPES)}
            isMultiple={true}
            allowClear={true}
            disabled={isDisabled}
            selected={this.props.selectedDocuments}
            onEnterPress={this.handleNewDocumentTypeAdded}
            selectAll={false}/>
    }

    getFooterButtonView(){
        if(_.isEmpty(this.props.selectedDocuments) || this.props.customerData.hasOwnProperty("clonedObject")){
            let content = _.isEmpty(this.props.selectedDocuments) ? "Please select at least one document type." :
                "There are unsaved changes in table. Please save or discard them first.";
            return <Popover content={content}>
                <Button disabled onClick={this.handleCreateButtonClicked}>Create</Button>
            </Popover>
        }

        return <Button onClick={this.handleCreateButtonClicked}>Create</Button>
    }

    render() {
        return (
            <div className={"classificationSelectionContainer"}>
                <div className={"dropDownsContainer"}>
                    <div className={"customerDropdown"}>
                        {this.getCustomerDropdownView()}
                    </div>
                    <Divider/>
                    <div className={"documentTypeDropdown"}>
                        {this.getDocumentDropdownView()}
                    </div>
                    <Divider/>
                </div>
                <div className={"buttonFooter"}>
                    {this.getFooterButtonView()}
                </div>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return state;
}
function mapDispatchToProps(dispatch) {
    let actions = bindActionCreators({ homeButtonClicked: myActions.handleScreenChanged });
    return { ...actions, dispatch };
}

const ConnectedView = connect(mapStateToProps, mapDispatchToProps)(ClassificationSelectionView);
export default ConnectedView
