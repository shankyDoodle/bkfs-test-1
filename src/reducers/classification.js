import {customerList, customerData} from "../mockdata/customerData";

export default {
    fetchCustomerList: function() {
        //implement server call here fro customer list
        return customerList;
    }
}