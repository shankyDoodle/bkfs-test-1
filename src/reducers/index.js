import getInitialState from "./initial-state";

import {
  HOME_BUTTON_CLICKED
} from '../actions'

function handleHomeButtonClicked(state, id) {
  console.log(state)
  return getInitialState()
}

export default (state = getInitialState(), action) => {
  switch (action.type) {
    case HOME_BUTTON_CLICKED:
      return handleHomeButtonClicked(state, action.id);

    default:
      return state
  }
}
