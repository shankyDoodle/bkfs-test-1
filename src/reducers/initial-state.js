import {screenNames} from "../constants/appConstants";

let initialState = {
    currentScreen: screenNames.HOME,
    customerList: {},
    selectedCustomers: [],
    documentTypes: {},
    selectedDocuments:[],
    customerData: {},
    classificationTableData: {},
    csvData:[],
    isScreenDirty: false,
    groupedDocumentElements:[],
    extractedSampleFile:null,
    textData:"",
    selectedDocumentSamples:{}
}

export const defaultState = initialState;

export const getInitialState = function () {
    return {
        ...initialState
    }
}