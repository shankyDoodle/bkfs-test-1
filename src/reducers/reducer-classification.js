import _ from 'lodash'

import {
    customerList,
    customerData,
    documentTypes
} from "../mockdata/mockData";

import * as appConstants from '../constants/appConstants';
import {Utils} from '../utils/utils';

export default {
    fetchCustomerData: function () {
        //implement server call here fro customer data
        return customerData;
    },

    setClassificationScreenOnLoadData: function(state, customerList, documentTypes){
        return {
            ...state,
            customerList,
            documentTypes
        }
    },

    handleClassificationDropDownOnBlur: function (state, dropdownType, selectedItems) {
        let oRet = {...state}
        if (dropdownType === appConstants.dropdownTypes.CUSTOMER_NAMES) {
            Object.assign(oRet, {selectedCustomers: selectedItems});
        } else {
            Object.assign(oRet, {selectedDocuments: selectedItems});
        }
        return oRet;
    },

    createCSVData: function(oTableData){
        let aData = [];

        let aHeaderData = oTableData.headerData;
        let aHeaderDataRow = [];
        for (let header of aHeaderData) {
            aHeaderDataRow.push(header.label);
        }
        aData.push(aHeaderDataRow);


        let aBodyData = oTableData.bodyData;
        _.forEach(aBodyData, function (oData) {
            let oRowData = oData.rowData;
            let row = [];
            for (let header of aHeaderData) {
                row.push(oRowData[header.label]);
            }
            aData.push(row);
        })

        return aData;
    },

    fetchSampleFiles: function(){
      //TODO: ideally fetch files from server
      return documentTypes;
    },

    handleClassificationCreateButtonCLicked: function (state, customerData, documentSamples) {
        let selectedDocuments = state.selectedDocuments;
        let selectedCustomers = state.selectedCustomers;

        let oCustomerList = state.customerList;
        let oDocumentTypes = state.documentTypes;

        let aHeaderData = [];
        aHeaderData.push({label: "Document Types"});
        for (let customer of selectedCustomers) {
            let label = oCustomerList[customer].label;
            aHeaderData.push({
                id: customer,
                label:label
            });
        }

        let aBodyData = [];
        for (let doc of selectedDocuments) {
            let sDocLabel = oDocumentTypes[doc].label;
            let values = {}
            values["Document Types"] = sDocLabel;

            for (let customer of selectedCustomers) {
                values[customer] = customerData[customer][sDocLabel]
            }
            let temp = {};
            temp["documentId"] = doc;
            temp["documentName"] = sDocLabel;
            temp["rowData"] = values
            temp["file"] = documentSamples[doc].data; //buffer data;
            temp["unusedFileData"] = documentSamples[doc];//backup
            aBodyData.push(temp);
        }


        let oTableData = {
            headerData: aHeaderData,
            bodyData: aBodyData
        }

        let aCSVData = this.createCSVData(oTableData);

        return {
            ...state,
            customerData,
            classificationTableData: oTableData,
            csvData: aCSVData
        }
    },


    handleTableCellDataChanged: function (state, customerId, docId, newVal) {

        //update local database
        let customerDataCloned = Utils.getDirtyData(state.customerData);
        customerDataCloned[customerId][docId] = newVal;

        //update table data
        let tableDataCloned = Utils.getDirtyData(state.classificationTableData);
        let bodyData = tableDataCloned.bodyData
        let editedRowIndex = _.findIndex(bodyData, function (o) {
            return o.documentId === docId;
        });
        let editedRowData = bodyData[editedRowIndex].rowData;
        editedRowData[customerId] = newVal;

        let aCSVData = this.createCSVData(tableDataCloned);

        return {
            ...state,
            isScreenDirty:true,
            csvData: aCSVData
        }
    },

    handleTableSaveDiscardClicked: function (state, buttonType) {
        let oRet = {...state}

        Object.assign(oRet, {isScreenDirty:false})
        if(buttonType === "save"){
            let customerData = state.customerData.clonedObject;
            let classificationTableData = state.classificationTableData.clonedObject;

            delete state.customerData.clonedObject
            delete state.customerData.isDirty
            delete state.classificationTableData.clonedObject;
            delete state.classificationTableData.isDirty;
            Object.assign(oRet, {customerData, classificationTableData});

            //TODO: handle any server calls to update Database
        }else {
            let customerData = state.customerData;
            let classificationTableData = state.classificationTableData;
            Object.assign(oRet, {customerData, classificationTableData});
        }

        delete state.customerData.clonedObject
        delete state.customerData.isDirty
        delete state.classificationTableData.clonedObject;
        delete state.classificationTableData.isDirty;

        return oRet;
    }

}