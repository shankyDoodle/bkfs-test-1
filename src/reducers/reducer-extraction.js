import _ from 'lodash'

import {
    documentTypes,
    groupedDocElements
} from "../mockdata/mockData";

import * as appConstants from '../constants/appConstants';
import {Utils} from '../utils/utils';

export default {
    fetchSampleFile: function (docName) {
        //TODO: ideally fetch files from server
        return "./pdfs/"+ documentTypes[docName];
    },

    fetchGroupedDocumentElementsByDocumentName: function (docName) {
        //TODO: ideally fetch files from server
        return groupedDocElements
    },

    setExtractionScreenOnLoadData: function (state, documentTypes) {
        return {
            ...state,
            documentTypes
        }
    },

    handleExtractionDropDownOnBlur: function (state, dropdownType, selectedItems) {
        return {
            ...state,
            selectedDocuments: [selectedItems]
        };
    },

    createSingleTextData: function (aGroupedDocumentData) {
        let sData = "";
        for (let oData of aGroupedDocumentData) {
            for (let dataElement of oData.dataElements) {
                sData+=dataElement+"\n";
            }
        }
        return sData;
    },

    handleExtractionCreateButtonCLicked: function (state) {
        let sSelectedDoc = state.selectedDocuments[0];
        let aGroupedDocumentData = this.fetchGroupedDocumentElementsByDocumentName(sSelectedDoc);
        let oFile = this.fetchSampleFile(sSelectedDoc);

        let textData = this.createSingleTextData(aGroupedDocumentData);

        return {
            ...state,
            textData,
            groupedDocumentElements: aGroupedDocumentData,
            extractedSampleFile: oFile
        }

    },

    moveWithinSameList: (list, startIndex, endIndex) => {
        const result = Array.from(list);
        const [removed] = result.splice(startIndex, 1);
        result.splice(endIndex, 0, removed);

        return result;
    },

    handleExtractionListDragEnd: function (state, source, destination) {
        let aGroups = Array.from(state.groupedDocumentElements);

        let sSourceGroupId = source.groupId
        let oFoundSourceGroup = aGroups.find(oG => oG.groupId === sSourceGroupId)
        let sDestGroupId = destination.groupId
        if (sSourceGroupId === sDestGroupId) {
            oFoundSourceGroup.dataElements = this.moveWithinSameList(oFoundSourceGroup.dataElements, source.index, destination.index);
        }else{
            let oFoundDestGroup = aGroups.find(oG => oG.groupId === sDestGroupId)
            let [removed] = oFoundSourceGroup.dataElements.splice(source.index, 1);
            oFoundDestGroup.dataElements.splice(destination.index, 0, removed);
        }

        let textData = this.createSingleTextData(aGroups);
        return {
            ...state,
            textData,
            groupedDocumentElements: aGroups
        }
    }
}