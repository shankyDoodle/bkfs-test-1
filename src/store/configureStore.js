import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { createLogger } from 'redux-logger'
// import api from '../middleware/api'
import rootReducer from '../reducers/reducer-index'
// import DevTools from '../containers/DevTools'

const configureStore = preloadedState => {
    const store = createStore(
        rootReducer,
        preloadedState,
        compose(
            applyMiddleware(thunk, createLogger())
        )
    )

    return store
}

export default configureStore
