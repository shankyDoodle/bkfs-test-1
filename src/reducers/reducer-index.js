import {getInitialState, defaultState} from "./initial-state";

import {screenNames} from "../constants/appConstants";
import * as appActions from '../actions'

import classification from "./reducer-classification";
import extraction from "./reducer-extraction";
import {HANDLE_EXTRACTION_DROP_DOWN_ON_BLUR, SET_EXTRACTION_SCREEN_ON_LOAD_DATA} from "../actions";



function resetToInitialState(oRet) {
    Object.assign(oRet, {...defaultState});
}
function handleScreenChanged(state, sScreenName) {
    let oRet = {...state}

    switch(sScreenName){
        case screenNames.HOME:
        default:
            resetToInitialState(oRet)
    }
    Object.assign(oRet, {currentScreen: sScreenName});
    return oRet
}

export default (state = getInitialState(), action) => {
    switch (action.type) {
        case appActions.HANDLE_SCREEN_CHANGED:
            return handleScreenChanged(state, action.screenName);

        case appActions.HANDLE_CLASSIFICATION_DROP_DOWN_ON_BLUR:
            return classification.handleClassificationDropDownOnBlur(state, action.dropdownButtonType, action.selectedItems);

        case appActions.HANDLE_CLASSIFICATION_CREATE_BUTTON_CLICKED:
            return classification.handleClassificationCreateButtonCLicked(state, action.customerData, action.documentSamples);

        case appActions.HANDLE_TABLE_CELL_DATA_CHANGED:
            return classification.handleTableCellDataChanged(state, action.customerName, action.docName, action.newVal);

        case appActions.HANDLE_TABLE_SAVE_DISCARD_CLICKED:
            return classification.handleTableSaveDiscardClicked(state, action.buttonType);

        case appActions.HANDLE_EXTRACTION_DROP_DOWN_ON_BLUR:
            return extraction.handleExtractionDropDownOnBlur(state, action.dropdownButtonType, action.selectedItems);

        case appActions.HANDLE_EXTRACTION_CREATE_BUTTON_CLICKED:
            return extraction.handleExtractionCreateButtonCLicked(state, action.groupedElements);

        case appActions.HANDLE_EXTRACTION_LIST_DRAG_END:
            return extraction.handleExtractionListDragEnd(state, action.source, action.destination);

        case appActions.SET_CLASSIFICATION_SCREEN_ON_LOAD_DATA:
            return classification.setClassificationScreenOnLoadData(state, action.customerList, action.documentTypes);

        case appActions.SET_EXTRACTION_SCREEN_ON_LOAD_DATA:
            return extraction.setExtractionScreenOnLoadData(state, action.documentTypes);

        default:
            return state
    }
}
