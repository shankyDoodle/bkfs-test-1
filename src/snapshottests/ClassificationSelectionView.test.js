import React from 'react';
import renderer from 'react-test-renderer';

import {Provider} from "react-redux";
import {getInitialState} from "../reducers/initial-state";
import configureStore from "../store/configureStore";

import ClassificationSelectionView from "../components/ClassificationSelectionView";

const state = getInitialState();
const store = configureStore(state);

const component = renderer.create(
    <Provider store={store}><ClassificationSelectionView /></Provider>,
);

test('Test ClassificationSelectionView component render', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
