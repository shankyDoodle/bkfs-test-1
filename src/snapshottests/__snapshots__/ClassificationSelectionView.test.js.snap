// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ClassificationSelectionView component render 1`] = `
<div
  className="classificationSelectionContainer"
>
  <div
    className="dropDownsContainer"
  >
    <div
      className="customerDropdown"
    >
      <div
        className="mssViewWrapper"
      >
        <div
          className="mssLabel"
        >
          Customer Names
        </div>
        <div
          className="ant-select ant-select-multiple ant-select-allow-clear ant-select-show-search"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <div
            className="ant-select-selector"
            onClick={[Function]}
            onMouseDown={[Function]}
          >
            <span
              className="ant-select-selection-search"
              style={
                Object {
                  "width": 0,
                }
              }
            >
              <input
                aria-activedescendant="undefined_list_0"
                aria-autocomplete="list"
                aria-controls="undefined_list"
                aria-haspopup="listbox"
                aria-owns="undefined_list"
                autoComplete="off"
                className="ant-select-selection-search-input"
                disabled={false}
                onChange={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                readOnly={true}
                role="combobox"
                style={
                  Object {
                    "opacity": 0,
                  }
                }
                value=""
              />
              <span
                aria-hidden={true}
                className="ant-select-selection-search-mirror"
              >
                
                 
              </span>
            </span>
            <span
              className="ant-select-selection-placeholder"
            >
              Please select
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-divider ant-divider-horizontal"
      role="separator"
    />
    <div
      className="documentTypeDropdown"
    >
      <div
        className="mssViewWrapper"
      >
        <div
          className="mssLabel"
        >
          Document Types
        </div>
        <div
          className="ant-select ant-select-multiple ant-select-allow-clear ant-select-show-search"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <div
            className="ant-select-selector"
            onClick={[Function]}
            onMouseDown={[Function]}
          >
            <span
              className="ant-select-selection-search"
              style={
                Object {
                  "width": 0,
                }
              }
            >
              <input
                aria-activedescendant="undefined_list_0"
                aria-autocomplete="list"
                aria-controls="undefined_list"
                aria-haspopup="listbox"
                aria-owns="undefined_list"
                autoComplete="off"
                className="ant-select-selection-search-input"
                disabled={false}
                onChange={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                readOnly={true}
                role="combobox"
                style={
                  Object {
                    "opacity": 0,
                  }
                }
                value=""
              />
              <span
                aria-hidden={true}
                className="ant-select-selection-search-mirror"
              >
                
                 
              </span>
            </span>
            <span
              className="ant-select-selection-placeholder"
            >
              Please select
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-divider ant-divider-horizontal"
      role="separator"
    />
  </div>
  <div
    className="buttonFooter"
  >
    <button
      className="ant-btn"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span>
        Create
      </span>
    </button>
  </div>
</div>
`;
